tests/flytekit/unit/core/test_async.py

```diff
@@ -0,0 +1,60 @@
+import pytest
+from flytekit.core.task import task, eager
+from flytekit.core.worker_queue import Controller
+from flytekit.utils.asyn import loop_manager
+from flytekit.core.context_manager import FlyteContextManager
+from flytekit.configuration import Config, DataConfig, S3Config, FastSerializationSettings, ImageConfig, SerializationSettings, Image
+from flytekit.core.data_persistence import FileAccessProvider
+from flytekit.tools.translator import get_serializable
+from collections import OrderedDict
+
+default_img = Image(name="default", fqn="test", tag="tag")
+serialization_settings = SerializationSettings(
+    project="project",
+    domain="domain",
+    version="version",
+    env=None,
+    image_config=ImageConfig(default_image=default_img, images=[default_img]),
+)
+
+
+@task
+def add_one(x: int) -> int:
+    return x + 1
+
+
+@eager(environment={"a": "b"})
+async def simple_eager_workflow(x: int) -> int:
+    # This is the normal way of calling tasks. Call normal tasks in an effectively async way by hanging and waiting for
+    # the result.
+    out = add_one(x=x)
+    return out
+
+
+@pytest.mark.asyncio
+async def test_easy_1():
+    res = await simple_eager_workflow(x=1)
+    print(res)
+    assert res == 2
+
+
+@pytest.mark.sandbox_test
+def test_easy_2():
```

Comment:
i think we should test 2 concurrent eager executions running in the same python vm but are independent

For example
exec-1 -> eager
exec-2 -> eager -> eager

Logically these are independent executions

Reply:
let me make an issue for this, and some of the other todos that came up.  will have to overhaul the flyte context & manager.  I think the scenario you're thinking of can be described like this.

```python
@task
def t1():
    print("normal task")

@eager
def eg_1():
    some_other_task()

@eager
def eg_p():
    t1()
    eg_1()
```

The issue is that the FlyteContext is a shared global across all coroutines (it's a thread local, not a coroutine local) which means that when t1's call handler runs, it'll set the execution state one way, and when eg_1's call handler runs it'll try to set it another way (actually it'll fail because it'll think it's running inside `t1` rather than `eg_p`).  What we want is a sort of tree of context objects rather than the list it is today.